lib/pg/connection.rb

```diff
@@ -71,13 +70,16 @@ def self::parse_connect_args( *args )
 		options.merge!( hash_arg )
 
 		if uri
-			uri.host     = nil if options[:host]
-			uri.port     = nil if options[:port]
-			uri.user     = nil if options[:user]
-			uri.password = nil if options[:password]
-			uri.path     = '' if options[:dbname]
-			uri.query    = URI.encode_www_form( options )
-			return uri.to_s.sub( /^#{uri.scheme}:(?!\/\/)/, "#{uri.scheme}://" )
+			query = URI.encode_www_form( options )
+			if uri.is_a? URI
+				uri = uri.to_s
+			end
+			if uri.include?("?")
```

Comment:
It looks like `?` does not definitely indicate there is a query. 

```
> URI('postgres://u?n:pw@localhost')              
=> #<URI::Generic postgres://u?n:pw@localhost>

> URI('postgres://un:p?w@localhost')
Traceback (most recent call last):
URI::InvalidURIError (bad URI(is not URI?): "postgres://un:p?w@localhost")
```

Reply:
That is interesting...and unfortunate. At the same time, I can't parse the URI that may have two hosts or more because that's definitely not valid. Hmm.